include/merlin/memory_pool.hpp

```diff
@@ -0,0 +1,600 @@
+struct StandardAllocator final : AllocatorBase<T, StandardAllocator<T>> {
+  using type = typename AllocatorBase<T, StandardAllocator<T>>::type;
+
+  static constexpr const char* name{"StandardAllocator"};
+
+  inline static type* alloc(size_t n, cudaStream_t ctx = nullptr) {
+    return new type[n];
+  }
+
+  inline static void free(type* ptr, cudaStream_t ctx = nullptr) {
+    delete[] ptr;
+  }
+};
+
+/**
+ * Claim/release buffers in pinned host memory.
+ */
+template <class T>
+struct HostAllocator final : AllocatorBase<T, HostAllocator<T>> {
+  using type = typename AllocatorBase<T, HostAllocator<T>>::type;
+
+  static constexpr const char* name{"HostAllocator"};
+
+  inline static type* alloc(size_t n, cudaStream_t ctx = nullptr) {
+    void* ptr;
+    CUDA_CHECK(cudaMallocHost(&ptr, sizeof(T) * n));
+    return reinterpret_cast<type*>(ptr);
+  }
+
+  inline static void free(type* ptr, cudaStream_t ctx = nullptr) {
+    CUDA_CHECK(cudaFreeHost(ptr));
+  }
+};
+
+/**
+ * Claim/release buffers in the active CUDA device. Will not test if the correct
+ * device was used, and throw if CUDA runtime API response is negative.
+ */
+template <class T>
+struct DeviceAllocator final : AllocatorBase<T, DeviceAllocator<T>> {
+  using type = typename AllocatorBase<T, DeviceAllocator<T>>::type;
+
+  static constexpr const char* name{"DeviceAllocator"};
+
+  inline static type* alloc(size_t n, cudaStream_t ctx = nullptr) {
+    void* ptr;
+    cudaError_t res;
+    if (ctx) {
+      res = cudaMallocAsync(&ptr, sizeof(T) * n, ctx);
+    } else {
+      res = cudaMalloc(&ptr, sizeof(T) * n);
+    }
+    CUDA_CHECK(res);
+    return reinterpret_cast<type*>(ptr);
+  }
+
+  inline static void free(type* ptr, cudaStream_t ctx = nullptr) {
+    cudaError_t res;
+    if (ctx) {
+      res = cudaFreeAsync(ptr, ctx);
+    } else {
+      res = cudaFree(ptr);
+    }
+    CUDA_CHECK(res);
+  }
+};
+
+/**
+ * Wrapper around another allocator that prints debug messages.
+ */
+template <class Allocator>
+struct DebugAllocator final
+    : AllocatorBase<typename Allocator::type, DebugAllocator<Allocator>> {
+  using type = typename Allocator::type;
+
+  static constexpr const char* name{"DebugAllocator"};
+
+  inline static type* alloc(size_t n, cudaStream_t ctx = nullptr) {
+    type* ptr = Allocator::alloc(n, ctx);
+    std::cout << Allocator::name << "[type_name = " << typeid(type).name()
+              << "]: " << static_cast<void*>(ptr) << " allocated = " << n
+              << " x " << sizeof(type) << " bytes, ctx = " << ctx << std::endl;
+    return ptr;
+  }
+
+  inline static void free(type* ptr, cudaStream_t ctx = nullptr) {
+    Allocator::free(ptr, ctx);
+    std::cout << Allocator::name << "[type_name = " << typeid(type).name()
+              << "]: " << static_cast<void*>(ptr) << " freed, ctx = " << ctx
+              << std::endl;
+  }
+};
+
+/**
+ * Helper structure to configure a memory pool.
+ */
+struct MemoryPoolOptions {
+  size_t buffer_size =
```

Comment:
What's the unit of the size? by byte or by elements?

Reply:
It is clarified in the comment! The comment says `size in bytes`.